File: 프로그래머스/3/72413. 합승 택시 요금/합승 택시 요금.py
import heapq
import sys

def solution(n, s, a, b, fares):
    INF = sys.maxsize
    answer = INF
    
    # s에서 출발해서 a, b로 택시를 탐
    # 합승을 하지 않아도 됨.
    # 둘이 겹치는 지점에서 하차, 각각 택시를 타서 계산함.
    # 돌아가지 않는 게 좋음. 사이클 x
    
    graph = [[] for _ in range(n+1)]
    
    # 그래프 초기화
    for fare in fares:
        node1,node2,cost = fare[0],fare[1],fare[2]
        graph[node1].append((node2,cost))
        graph[node2].append((node1,cost))
        
    
    # 다익스트라
    def dijkstra(start):
        pq = []
        dist = [INF] * (n+1)
        heapq.heappush(pq,(0,start))
        dist[start] = 0
        
        while(pq):
            cost, node = heapq.heappop(pq)
            
            for node in graph[node]:
                nextNode = node[0]
                nextCost = cost + node[1]
                
                if (nextCost > dist[node[0]]):
                    # 이미 최소임
                    continue
                
                dist[nextNode] = nextCost
                heapq.heappush(pq, (nextCost, nextNode))
        
        return dist
    
    dist_s = dijkstra(s)
    dist_a = dijkstra(a)
    dist_b = dijkstra(b)
    
        
    for k in range(1,n+1):
        # 중간 지점 k

        answer = min(answer, dist_s[k] + dist_a[k] + dist_b[k])
        
    
    return answer

File: 프로그래머스/3/72413. 합승 택시 요금/합승 택시 요금.js
function solution(n, s, a, b, fares) {
    var answer = Infinity;
    
    // 1. 다익스트라: 시작-환승지점 +환승지점-a + 환승지점-b 근데 그러면 다익스트라 두 번만 하면 되나?
    // 2. 플로이드 워셜: 전체 노드 간 최단 거리를 구하는 것이므로 O(n^3)이고, ... 어케 풀지?
    
    // fares 그래프로 변환
    const graph = Array.from({length:n+1}, () => Array(n+1).fill(Infinity))
    
    for (const [x,y,dist] of fares) {
        graph[x][y] = Math.min(dist,graph[x][y])
        graph[y][x] = Math.min(dist, graph[y][x])
    }
    
    function dijkstra(node) {
        
        const dist = Array(n+1).fill(Infinity)
        const visit = Array(n+1).fill(0)
        dist[node] = 0

        
        for (let i=1; i<=n; i++) {
               
            let current = -1
            let minDist = Infinity
            
            for (let j=1; j<=n; j++) {
                
                if (visit[j] !== 1 && dist[j] < minDist) {
                    minDist = dist[j]
                    current = j
                }
                
            }
          
        
        if (current === -1) break
        
        visit[current] = 1
        
         for (let next=1; next<=n; next++) {
            
             if (visit[next] === 1) continue
             if (graph[current][next] === Infinity) continue
             
             nextDist = dist[current] + graph[current][next]
             
             if (nextDist < dist[next]) {
                 dist[next] = nextDist
             }
             
         }
        }

        return dist 
    }
    const distS = dijkstra(s)
    const distA = dijkstra(a)
    const distB = dijkstra(b)
    
    
    // 합승 지점 k
    for (let k=1; k <= n; k++) {
        answer = Math.min(answer, distS[k] + distA[k] + distB[k])
    }
    
    
    
    return answer;
}